### PR TITLE
Add TokenRequestOptions and ClientSecretCredentialOptions

### LINKED ISSUE #1176: Token Credential: add options (GetTokenOptions), add Transport there, move authority there
+ aet rid of multiple constructors taking scopes in different formats.

This issue is a change request from the API review: https://apiview.dev/Assemblies/Review/6f5301c273ee4eda8d35487a898439a6

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #1176.

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/credentials.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials.hpp:41]: Why are we dong this?

File: sdk/core/azure-core/inc/azure/core/credentials.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials.hpp:49]: Need docs

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/policy.hpp:332]:
```suggestion
        GetTokenOptions const& options)
```
Shouldn't this be optional?

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/policy.hpp:337]: Is this static cast safe? What exactly are we trying to do here with the clone/release of the transport policy passed in? Could this be a `std::move` instead?

File: sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp:33]: Does this need to be mandatory?

**Comment by antkmsft** [comment]
@Jinming-Hu, you want your team to review changes like this, please review, we want this to hit `master` on Monday.

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:332]: I don't think it is useful without scopes.
Optionality rules should not apply the same way - it is not options for the policy, these are options for GetToken() operation.

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:337]: It is safe. We have to have it. `Clone()` returns `unique_ptr<HttpPolicy>`. The struct needs `unique_ptr<TransportPolicy>`. But `TransportPolicy->Clone()` still returns `unique_ptr<HttpPolicy>`. `TransportPolicy` derives from `HttpPolicy`. `static_cast` would not compile if they are unrelated. There is no cast from `unique_ptr<Base>` to `unique_ptr<Derived>`. So, you have to do `unique_ptr<Base>.release()` to get `Base*` (and `release()` tells `unique_ptr<Base>` to release control of it - it returns the original ptr, puts null on its place, so it would not dealloc the ptr in `~unique_ptr<Base>()` destructor). `reset()` takes ownership of `T*` (in this case, `Derived*` which we get by casting `static_cast<Derived*>(basePtr)`). As a result, we end up with the `unique_ptr<Derived>`.
Even if we didn't have to do the cast thing, we cannot use `move()`, because that would require to take the `options` by value, i.e. as a copy, but `unique_ptr`s are non-copyable, because they have exclusive ownership, and you can't (reasonably) have two `unique_ptr`s pointing at the same object.

File: sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp
Comment (antkmsft) [reply sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp:33]: I think it is better to have it mandatory. Tecnically, you don't have to initialize it, but then it will be null, and nothing good would happen. We can always add a logic that if `GetTokenOptions::TransportPolicy` is null, then use `Details::GetTransportAdapter()`, but I don't think it is good if client libraries would do that. It is up to client library to put a non-null value there, the same way that they do it for their own `TransportPolicy`. So, this does not prevent client libs from using the default - either here or for their own `TransportPolicyOptions`, we're just saying don't give us null, initialize the at least the same way you initialize your own `TransportPolicyOptions`, do it consciously, or we may end up with the clients that have their own HTTP transport customizable, but the transport for authentication would always be using the default value, and that would be a miss.

File: sdk/core/azure-core/inc/azure/core/credentials.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/credentials.hpp:41]: It is a forward declaration to let the compiler know that whenever it sees `Http::TransportPolicy`, it is a class named `TransportPolicy` inside the `Http` namespace. We can't include `policy.hpp` here, because `policy.hpp` needs `credentials.hpp`, since `BearerTokenAuthenticationPolicy` does contain `GetTokenOptions` by value a a member, and that requires the full `GetTokenOptions` declaration to be known to the compiler, not just the mere knowledge that it is a class - it needs to know all the members, because it needs to know the `sizeof(GetTokenOptions)` in order for it to be a class member contained by value. `GetTokenOptions`, on the other hand, is fine just knowing that `Http::TransportPolicy`, it is a class named `TransportPolicy` inside the `Http` namespace, because, yes, it is used as a data member of `GetTokenOptions`, but it is contained as a pointer, and for that, you don't need the `sizeof(Http::TransportPolicy)` - compiler is using a `sizeof(pointer)`, and all pointers (to data types) are of the same size, be it `int*` or `Http::TransportPolicy*`.

File: sdk/core/azure-core/inc/azure/core/credentials.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials.hpp:70]: Add changelog entry for breaking and new API changes.

**Comment by antkmsft** [comment]
@ahsonkhan, I rethought some of the options, so some of your former feedback no longer applies - in a good way, I believe.

**Comment by antkmsft** [comment]
@RickWinter, @vhvb1989, @ahsonkhan, @Jinming-Hu, if we want this for Tuesday, I need reviews and approvals.

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (vhvb1989) [inline sdk/core/azure-core/inc/azure/core/http/policy.hpp:343]: Wonder if getTokenOptions should be optional and have a default init just like we did on other policies like the transport and telemetry...

```suggestion
        GetTokenOptions getTokenOptions = GetTokenOptions ())
```

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/policy.hpp:312]:
```suggestion
  struct TokenRequestOptions
```

Note that other languages call this struct "TokenRequestContext".  I don't like "Context", though we may want to match that name to match what other language SDKs are doing

https://docs.microsoft.com/en-us/dotnet/api/azure.core.tokenrequestcontext?view=azure-dotnet
https://docs.microsoft.com/en-us/java/api/com.azure.identity.chainedtokencredential.gettoken?view=azure-java-stable

File: sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp:19]: why the 2 minutes?  Looks like we need a comment here about it.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp
Comment (vhvb1989) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:39]: should this be in ::details?

File: sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp:33]: Would this work: ???

```suggestion
    policies.emplace_back(
        std::make_unique<BearerTokenAuthenticationPolicy>(credential, {{"https://vault.azure.net/.default"}}));
```

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp
Comment (RickWinter) [inline sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:71]: Per our guidelines this should be a ClientSecretCredentialOptions and the ClientSecretCredentialsOptions would derive from TokenCredentialOptions

**Review by RickWinter — CHANGES_REQUESTED** [review]

**Review by vhvb1989 — COMMENTED** [review]
Some questions/comments. We will need @Jinming-Hu  and @katmsft approval for this one

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:343]: +1, The default for options parameters follows the C++ design guidelines

File: sdk/identity/azure-identity/src/client_secret_credential.cpp
Comment (RickWinter) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:39]: Does it need to be global or can it be a static member in the TokenCredentialOptions class?

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:343]: This was brought up earlier: https://github.com/Azure/azure-sdk-for-cpp/pull/1527#discussion_r567172776

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:312]: I actually like the "TokenRequestContext/CancellationToken", and I like "TokenRequestOptions", but this is what was decided on during the review meeting.
I would also be ok with "BearerTokenAuthenticationPolicyOptions", and I am not a fan of a verb in the class name. But also we had that time to discuss, all of us were here, and we decided on GetTokenOptions, that the only thing that makes me hesitate.

@JeffreyRichter, @schaabs, please let me know if you would still prefer GetTokenOptions name instead of TokenRequestOptions.

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp
Comment (antkmsft) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:71]: This is not how .NET has it, but I agree, ClientSecretCredentialOptions makes sense. I would not have TokenCredentialOptions for now then, because ClientSecretCredential is the only user so far. If needed, we can add TokenCredentialOptions later, and make ClientSecretCredentialOptions inherit it, without breaking change.

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:343]: The reason I did not follow the pattern is that GetTokenOptions has scopes. A policy without scopes set is useless.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:39]: static is global, but with class/struct scope, so syntax is a git different. It is private static, so it would not be a breaking change to move it to Details.

File: sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp
Comment (antkmsft) [reply sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp:33]: Did not compile on MSVC for me :(

File: sdk/identity/azure-identity/src/client_secret_credential.cpp
Comment (RickWinter) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:39]: I was referring to the g_aad* name.  Our guidelines are g_ is for globals.  This is not a global so it shouldn't be g_

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp
Comment (ahsonkhan) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:71]: > and make ClientSecretCredentialOptions inherit it, without breaking change.

I agree it won't be breaking. I wish we had a reference doc with some basic examples of when things are considered breaking or not, specifically in C++. @CaseyCarter - do you have a good reference?

File: sdk/identity/azure-identity/src/client_secret_credential.cpp
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:39]: Statics are globals, we had that discussion. Prefixes for statics are exactly the same, `g_`.

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp
Comment (antkmsft) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:71]: I think if the old code still compiles when you compile the new SDK lib, and your code, it is not considered breaking; not the ABI.

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:343]: Not fixing this, please reply if you insist, let's talk.

**Comment by antkmsft** [comment]
@RickWinter, please unblock if you like it now.

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp]: Is it there is no valid set of default scopes we could set for scope?

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp]: Scopes currently used do look like this: `https://storage.azure.com/.default`, `https://vault.azure.net/.default`.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp]: I.e. scope does include service name: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#scopes-and-permissions

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp]: Also, from https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#the-default-scope: "The scope is constructed by using the resource URI and /.default." So, it is always some resource URI, I see no reasonable default.

Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp]: Thanks for the link.  That implies there are mandatory parameters to create the token options.  Not so keen on the name being options since they are required parameters....  
cc: @schaabs

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp]: `TokenRequestParameters`? `TokenRequestContext`?
I am not fluent enough in English, does "options" that heavy imply things being non-mandatory, or is it more about following the pattern that is seen in other places across the SDK?

**Pull request merged**
